src/Serializer/SnippetDenormalizer.php

```diff
@@ -0,0 +1,51 @@
+<?php
+
+namespace eLife\ApiSdk\Serializer;
+
+use eLife\ApiSdk\Promise\CallbackPromise;
+use GuzzleHttp\Promise\PromiseInterface;
+
+final class SnippetDenormalizer
+{
+    private $id;
+    private $fetcher;
+    private $identityMap;
+    private $globalCallback;
+
+    public function __construct(callable $id, callable $fetcher)
+    {
+        $this->id = $id;
+        $this->fetcher = $fetcher;
+        $this->identityMap = new IdentityMap();
+    }
+
+    public function denormalizeSnippet(array $item) : PromiseInterface
+    {
+        $id = call_user_func($this->id, $item);
+
+        if ($this->identityMap->has($id)) {
```

Comment:
We do need a cache here otherwise we can't do parallel lazy loading. It doesn't store much in memory until something triggers it, then anything that's ever been recorded will be requested. This includes models that have been forgotten about.